code/ecdsap256sha2/Hacl.Spec.ECDSA.fst

```diff
@@ -0,0 +1,314 @@
+    if uint_v i = 0 then pNew == p /\ qNew == q
+    else
+      let p1, q1 = swap p q in 
+      p1 == pNew /\ q1 == qNew
+ }
+)
+
+#reset-options "--z3refresh --z3rlimit  100"
+
+let conditional_swap i p q = 
+  if uint_v i = 0 then 
+    (p, q)
+  else
+    (q, p)
+
+val lemma_swaped_steps: p: nat_prime -> q: nat_prime -> 
+  Lemma (
+    let (afterSwapP, afterSwapQ) = swap p q in 
+    let p1, q1 = _exp_step0 afterSwapP afterSwapQ in 
+    let p2, q2 = swap p1 q1 in 
+    let r0, r1 = _exp_step1 p q in 
+    p2 == r0 /\ q2 == r1)
+
+let lemma_swaped_steps p q = ()
+
+open Lib.RawIntTypes 
+
+val _exp_step: k:  lseq uint8 32 ->  i: nat{i < 256} -> tuple2 nat_prime nat_prime ->
+  Tot (t: tuple2 nat_prime nat_prime)
+
+let _exp_step k i (p, q) = 
+  let bit = 255 - i in 
+  let bit = ith_bit k bit in 
+  if uint_to_nat bit = 0 then 
+      _exp_step0 p q 
+  else _exp_step1 p q  
+
+#reset-options "--z3refresh --z3rlimit 300"
+
+val _exponent_spec: k: lseq uint8 32  -> tuple2 nat_prime nat_prime -> Tot (tuple2 nat_prime nat_prime)
+
+let _exponent_spec k (p, q) = 
+  Lib.LoopCombinators.repeati 256 (_exp_step k) (p, q)
+
+
+unfold let prime_p256_order_inverse_list: list uint8 = 
+   [
+     (u8 79); (u8 37); (u8 99); (u8 252); (u8 194); (u8 202); (u8 185); (u8 243); 
+     (u8 132); (u8 158); (u8 23); (u8 167); (u8 173); (u8 250); (u8 230); (u8 188); 
+     (u8 255); (u8 255); (u8 255); (u8 255); (u8 255); (u8 255); (u8 255); (u8 255);
+     (u8 0); (u8 0); (u8 0); (u8 0); (u8 255); (u8 255); (u8 255); (u8 255)
+   ]
+
+let prime_p256_order_inverse_seq: lseq uint8 32 = 
+  assert_norm (List.Tot.length prime_p256_order_inverse_list == 32);
+  of_list prime_p256_order_inverse_list
+
+
+unfold let prime_p256_order_list: list uint8 = 
+   [
+     (u8 81); (u8 37); (u8 99); (u8 252); (u8 194); (u8 202); (u8 185); (u8 243); 
+     (u8 132); (u8 158); (u8 23); (u8 167); (u8 173); (u8 250); (u8 230); (u8 188); 
+     (u8 255); (u8 255); (u8 255); (u8 255); (u8 255); (u8 255); (u8 255); (u8 255);
+     (u8 0); (u8 0); (u8 0); (u8 0); (u8 255); (u8 255); (u8 255); (u8 255)
+   ]
+
+let prime_p256_order_seq: lseq uint8 32 = 
+  assert_norm (List.Tot.length prime_p256_order_list == 32);
+  of_list prime_p256_order_list
+
+open Hacl.Spec.P256.Definitions
+
+val exponent_spec: a: nat_prime -> Tot (r: nat_prime {r = pow a (prime_p256_order - 2) % prime_p256_order})
+
+let exponent_spec a = 
+    let a0, _ = _exponent_spec prime_p256_order_inverse_seq (1, a) in
+    admit();
+    a0
+
+
+val changeEndian: i: felem_seq -> Tot felem_seq
+
+let changeEndian i = 
+  let zero =  Lib.Sequence.index i 0 in 
+  let one =   Lib.Sequence.index i 1 in 
+  let two =   Lib.Sequence.index i 2 in 
+  let three = Lib.Sequence.index i 3 in 
+
+  let o = Lib.Sequence.upd i 0 three in 
+  let o = Lib.Sequence.upd o 1 two in 
+  let o = Lib.Sequence.upd o 2 one in 
+	  Lib.Sequence.upd o 3 zero
+
+
+
+val verifyQValidCurvePointSpec: publicKey: tuple3 nat nat nat -> bool
```

Comment:
The verification wouldn't be complete without also checking `~(isPointAtInfinity publicKey)`. 
This function is currently used at places where this holds (following a translation of `publicKey` to Jacobian coordinates), so adding this as a pre-condition suffices.

Note that the corresponding implementation of this specification includes this pre-condition:
https://github.com/project-everest/hacl-star/blob/bb21cf455b1e6242fe0079dc56cc85397098b8b9/code/ecdsap256sha2/Hacl.Impl.P256.Signature.Common.fst#L249

Reply:
Done.